[BottomSheet] Add BottomSheet component

antialiasis opened the pull request:
Hey, first time contributing. I needed a BottomSheet component for a project, and it seemed appropriate to submit it back to Material-UI.

It follows [the spec](https://www.google.com/design/spec/components/bottom-sheets.html) reasonably closely, except for the part about the sheet not initially expanding beyond the 16:9 keyline on mobile - as far as I'm aware material-ui doesn't really do specialized mobile behaviour as it is and has no facilities to support that, although since I'm not _that_ familiar with the entire codebase I may be mistaken.

Following that, I was unsure how to deal with the possibility in the spec of fully expanding a bottom sheet - the way it's described there, the sheet starts out at the 16:9 keyline but can be expanded by swiping it up, but if it's not starting at the 16:9 keyline, that doesn't make sense (either the content is small enough to fit, in which case the sheet'll fit the content and shouldn't be expanded, or the content is bigger than the screen, in which case the sheet's max-height will take effect in accordance with another part of the spec, making it effectively fully expanded from the start).

If you have any thoughts on doing this better, I'm all for improving the pull request.

Includes docs demo, but no tests - this is heavily based on the Drawer component, which has no tests, and I wasn't sure exactly what you would want tested here, if anything - I can't tell if the slash in that checklist item means "and" or "or".

Closes #2126 (although see comments above on fully expanded sheets).

- [x] PR has tests / docs demo, and is linted.
- [x] Commit and PR titles begin with [ComponentName], and are in imperative form: "[Component] Fix leaky abstraction".
- [x] Description explains the issue / use-case resolved, and auto-closes the related issue(s) (http://tr.im/vFqem).

fannarsh commented:
@oliviertassinari and @mbrookes any comments on this PR?

mbrookes commented:
@antialiasis Thanks for the contribution. Did you consider making this a feature of Drawer?

antialiasis commented:
Hi, thanks for responding.

While this component is based on the Drawer and has many similar features, there are also significant differences. For example, while a drawer has a set width, according to the spec a bottom sheet should adapt to the height of its content; this necessitates a somewhat different approach to how we keep it shifted out of frame while closed. Bottom sheets should be closeable by sliding down, like drawers are closeable by sliding towards the edge, but unlike with drawers, we can't generally decide whether the user is trying to scroll or slide based on the direction they move in but rather must decide based on where the touch originates. Some drawer defaults just don't make sense for bottom sheets: as far as I can tell from the spec, bottom sheets are never supposed to open when sliding up from the bottom, but to accomplish this with a drawer the user needs to explicitly set disableSlideToOpen, and following from that, it doesn't really make sense for bottom sheets to be uncontrolled, either. It'd be _possible_ to combine the two sensibly to make a generic drawer component that can open from any side of the screen, but it'd call for a lot of probably backwards-incompatible changes to the Drawer component and involve quite a bit of specialized behaviour based on whether it's horizontal or vertical. Perhaps it'd make sense to do that anyway, but all in all a new component seemed like the more conservative approach, compared to cutting up an existing one, introducing backwards-incompatibilities and possibly breaking existing behaviour.

I would also say that drawers and bottom sheets, while similar in the sense that they slide in from an edge of the screen and can be closed by sliding back, are conceptually pretty different and made for different use cases. Drawers are navigation menus and a page generally has exactly one of them (_possibly_ two); they are top-level objects. Bottom sheets are either freeform extra content/information/controls supplementing the main view or essentially a mobile alternative to drop-down menus or dialogs; they're fundamentally context-sensitive. This is why they call for different defaults.

It's true the bottom sheet component repeats a lot of Drawer's functionality, though. Perhaps it'd be ideal to have a generic component implementing the slide-in/out functionality from any direction and have both Drawer and BottomSheet make use of that.

mbrookes commented:
> Perhaps it'd be ideal to have a generic component implementing the slide-in/out functionality from any direction and have both Drawer and BottomSheet make use of that.

That sounds like a sensible approach, otherwise we'll find ourself maintaining multiple variations of the same code in different components. [Bottom Nav](https://www.google.com/design/spec/components/bottom-navigation.html) could also potentially take advantage in the future.

nathanmarks commented:
@antialiasis 100% about the functionality repetition.

Yes, when it comes down to specifics they are different. However, they share a significant amount of underlying concepts/functionality and I believe we'd be better off finding a design that promotes more code sharing in this situation. 

I'd love to hear/see any ideas you have.

antialiasis commented:
I'll do the proposed refactor of the sliding functionality into a separate single component that can be used by both Drawer and BottomSheet and possible future components, then, and you can see what you think.

antialiasis commented:
Hello again. I've refactored the shared drawer/bottom sheet functionality into a shared internal component, SlidingSheet. I believe everything should work as before in both Drawer and BottomSheet.

oliviertassinari commented:
@antialiasis Interesting PR, I'm gonna have a look at it. Thanks for working on this!

oliviertassinari commented on docs/src/app/components/pages/components/BottomSheet/ExampleCloseable.js:
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
```
```js
import React, {Component} from 'react';
```

oliviertassinari commented on docs/src/app/components/pages/components/BottomSheet/ExampleCloseable.js:
```diff
@@ -0,0 +1,31 @@
+export default class BottomSheetCloseableExample extends React.Component {
+
+  constructor(props) {
+    super(props);
```
We don't provide any properties. What about?
```js
super();
```

oliviertassinari commented on docs/src/app/components/pages/components/BottomSheet/ExampleCloseable.js:
```diff
@@ -0,0 +1,31 @@
+import MenuItem from 'material-ui/MenuItem';
+import RaisedButton from 'material-ui/RaisedButton';
+
+export default class BottomSheetCloseableExample extends React.Component {
```
We use the following convention in our other examples: `BottomSheetExampleCloseable`.

oliviertassinari commented on docs/src/app/components/pages/components/BottomSheet/Page.js:
```diff
@@ -0,0 +1,66 @@
+import bottomSheetModalExampleCode from '!raw!./ExampleModal';
+import bottomSheetCode from '!raw!material-ui/BottomSheet/BottomSheet';
+
+const descriptions = {
```
Using this pattern is deprecated. We can simply add the description above the example.
For instance, have a look at https://github.com/callemall/material-ui/blob/master/docs/src/app/components/pages/components/Dialog/ExampleAlert.js#L7.

oliviertassinari commented on src/BottomSheet/BottomSheet.js:
```diff
@@ -0,0 +1,147 @@
+import React, {Component, PropTypes} from 'react';
+import ReactDOM from 'react-dom';
```
What about?
```js
import {findDOMNode} from 'react-dom';`
```

oliviertassinari commented on src/BottomSheet/BottomSheet.js:
```diff
@@ -0,0 +1,147 @@
+class BottomSheet extends Component {
+  static propTypes = {
+    /**
+     * The contents of the `BottomSheet`
```
missing dot at the end of the sentence 😁 .

oliviertassinari commented on src/BottomSheet/BottomSheet.js:
```diff
@@ -0,0 +1,147 @@
+     * full-width.
+     */
+    width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
+
```
We can kill this line 😁.

oliviertassinari commented on src/BottomSheet/BottomSheet.js:
```diff
@@ -0,0 +1,147 @@
+
+  componentDidMount() {
+    this.updateHeight();
+    window.addEventListener('resize', this.updateHeight);
```
What about using this pattern https://github.com/callemall/material-ui/blob/master/src/Popover/Popover.js#L355?

oliviertassinari replied:
Well, [`withWidth`](https://github.com/callemall/material-ui/blob/master/src/utils/withWidth.js#L61) is a better example.

oliviertassinari commented on src/BottomSheet/BottomSheet.js:
```diff
@@ -0,0 +1,147 @@
+      overlayStyle,
+      style,
+      width,
+    } = this.props;
```
Should we spread the rest of the properties to the root element?

oliviertassinari commented on src/BottomSheet/BottomSheet.js:
```diff
@@ -0,0 +1,147 @@
+
+    return (
+      <SlidingSheet
+        ref="sheet"
```
*string* refs are kind of deprecated by React. Still, I'm fine keeping it here as we do in all of our components.

oliviertassinari commented on src/BottomSheet/index.js:
```diff
@@ -0,0 +1 @@
+export default from './BottomSheet';
```
We should probably export it in the `src/index.js` file too.

oliviertassinari commented on src/Drawer/Drawer.js:
```diff
@@ -344,48 +142,36 @@ class Drawer extends Component {
       style,
+      swipeAreaWidth,
       zDepth,
     } = this.props;
```
Should we spread the rest of the properties to the root element?

oliviertassinari commented on src/Drawer/Drawer.js:
```diff
@@ -344,48 +142,36 @@ class Drawer extends Component {
+        open={this.state.open}
+        overlayClassName={overlayClassName}
+        overlayStyle={overlayStyle}
+        primaryDimension={this.props.width || this.context.muiTheme.drawer.width}
```
Why not destructuring the `width` too?

oliviertassinari commented:
@antialiasis I will continue reviewing this PR tomorrow. Great start 👍 .
It would be good to rebase on top of master, I had to run `npm install`.

oliviertassinari commented on src/BottomSheet/BottomSheet.js:
```diff
@@ -0,0 +1,147 @@
+
+  close(reason) {
+    if (this.props.onRequestClose) this.props.onRequestClose(reason);
+    return this;
```
Why returning `this`? `close` is a private API, we don't need to chain the calls.

oliviertassinari commented on src/internal/SlidingSheet.js:
```diff
@@ -0,0 +1,439 @@
+  };
+
+  componentWillMount() {
+    this.maybeSwiping = false;
```
Dead property?

oliviertassinari commented on src/internal/SlidingSheet.js:
```diff
@@ -0,0 +1,439 @@
+
+  componentWillMount() {
+    this.maybeSwiping = false;
+    this.touchStartX = null;
```
Dead property?

oliviertassinari commented on src/Drawer/Drawer.js:
```diff
@@ -107,14 +99,9 @@ class Drawer extends Component {
 
```
The above block is no longer usefull:
```js
     this.maybeSwiping = false;
     this.touchStartX = null;
     this.touchStartY = null;
     this.swipeStartX = null;
```

oliviertassinari commented on src/internal/SlidingSheet.js:
```diff
@@ -0,0 +1,439 @@
+class SlidingSheet extends Component {
+  static propTypes = {
+    /**
+     * The contents of the `SlidingSheet`
```
Missing dot at the end of the sentence.

oliviertassinari commented on src/internal/SlidingSheet.js:
```diff
@@ -0,0 +1,439 @@
+      overlayStyle,
+      style,
+      zDepth,
+    } = this.props;
```
Same here. I would spread the other properties to the root component.
Who knows the need from of our users. Let's give them flexibility.

oliviertassinari commented on src/internal/SlidingSheet.js:
```diff
@@ -0,0 +1,439 @@
+    return (
+      <div
+        className={className}
+        style={Object.assign({width: '100%'}, style)}
```
Should we move `{width: '100%'}` into `getStyles`?

oliviertassinari replied:
Would be also great to use `prepareStyles`.

oliviertassinari commented on src/internal/SlidingSheet.js:
```diff
@@ -0,0 +1,439 @@
+    const width = vertical ? '100%' : this.props.primaryDimension;
+
+    const styles = {
+      root: {
```
this style is not applied on the `root`. We could use a different name, like paper?

oliviertassinari commented on src/internal/SlidingSheet.js:
```diff
@@ -0,0 +1,439 @@
+import React, {Component, PropTypes} from 'react';
+import ReactDOM from 'react-dom';
```
```js
import {findDOMNode} from 'react-dom';
```

oliviertassinari commented on src/internal/SlidingSheet.js:
```diff
@@ -0,0 +1,439 @@
+          ref="clickAwayableElement"
+          zDepth={zDepth}
+          rounded={false}
+          transitionEnabled={!this.state.swiping}
```
I would say that we don't need this. The style is quite static.

antialiasis replied:
Isn't the point of disabling transitions when swiping that while we're swiping we want to manually set the height when the user moves their finger, and transitions would create a delay between the user's action and the UI responding? Or are you referring to something else?

oliviertassinari replied:
Good point. I have overlooked that we use the ref `clickAwayableElement` to change the `transform` property 👍 .

oliviertassinari commented on src/internal/SlidingSheet.js:
```diff
@@ -0,0 +1,439 @@
+    document.body.addEventListener('touchcancel', this.onBodyTouchEnd);
+  };
+
+  setPosition(translatePrimary) {
```
*Side comment that is out of the scope of this PR*:
Notice that `setPosition` can potentially be triggered at a higher frequency than the screen refresh rate (60Hz).
See https://github.com/callemall/material-ui/issues/2862 as reference.

antialiasis commented:
Hello and thanks for the review! Apologies for not getting back to you earlier; I've been on vacation. I'll take a stab at improving the PR according to your comments soon.

oliviertassinari commented:
@antialiasis Great, thanks!

antialiasis commented:
Hello again! I've made the mentioned improvements to the PR. Since most of the issues mentioned were actually inherited from Drawer (should have made sure to familiarize myself with current practices!), and the PR makes changes to Drawer anyway, I fixed them there as well; also rebased the whole thing off master, as suggested.

oliviertassinari commented:
@antialiasis That's a great contribution! Thanks a lot. I'm have noticed more stuff that we could improve on this PR. I'm gonna carry it to master. Expect a new PR soon starting from your commit.

skirunman commented:
I saw updated roadmap and plan was to minimize new features in 0.15.3. However, this one plus support for bottom navigation are key for our mobile app https://github.com/callemall/material-ui/issues/3712 

Would be awesome to have these features in the next release so we at least have something to work with before the major refactor release 0.16 is available in a few months. Thanks for the consideration.

oliviertassinari commented:
I'm closing this PR as I have based #4732 on top of this one. It would be better to keep one place to get feedbacks. 
Thanks for this contribution.
